### PR TITLE
Isolate benchmarking invoke + statistic tests

### PR DESCRIPTION
**Pull request by Arcticae**

### CODE REVIEW

File: tests/performance/test_performance_testing.py [hunk 1]
```diff
@@ -97,44 +102,112 @@ def build_test_suite(
...
 
-        assert (
-            not tests_state.any_failed_or_broken()
-        ), "Tests failed! See logs above for more info"
+        wait_for_completion(test_suite, shared_state)
 
 
-def test_deploy_perf(benchmark, tmp_path, basic_contract_path):
+@pytest.mark.asyncio
+async def test_deploy_perf(aio_benchmark, tmp_path, basic_contract_path):
```
Comment (kasperski95) [inline tests/performance/test_performance_testing.py:210]: Marking tests with `@pytest.mark.asyncio` is unnecessary thanks to the following config:

pytest.ini
```
[pytest]
asyncio_mode=auto
```

File: tests/performance/test_performance_testing.py [hunk 2]
```diff
@@ -97,44 +102,112 @@ def build_test_suite(
...
+        tests_left -= 1
+
+    assert (
+        not tests_state.any_failed_or_broken()
+    ), "Tests failed! See logs above for more info"
+
+
+# https://github.com/ionelmc/pytest-benchmark/issues/66#issuecomment-575853801
+@pytest.fixture(scope="function", name="aio_benchmark")
+def aio_benchmark_fixture(benchmark):
```
Comment (kasperski95) [inline tests/performance/test_performance_testing.py:144]: `aio_` — What does it mean?

File: tests/performance/test_performance_testing.py [hunk 3]
```diff
@@ -97,44 +102,112 @@ def build_test_suite(
...
+def wait_for_completion(test_suite: TestSuite, tests_state: SharedTestsState):
+    tests_left = len(test_suite.test_case_names)
+    while tests_left:
+        test_result = tests_state.get_result()
+        if isinstance(test_result, PassedTestCase):
+            out = sys.stdout
+        else:
+            out = sys.stderr
+
+        print(test_result.format(), file=out)
```
Comment (kasperski95) [inline tests/performance/test_performance_testing.py:134]: Is `print` needed here?

File: tests/performance/statistical_test.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def extract_measure(before, after, test_name):
+    before_test = find_test(before["benchmarks"], test_name)
+    after_test = find_test(after["benchmarks"], test_name)
+    return calc_t_student_measure(before_test, after_test)
+
+
+def assert_n(run, n):
+    """This functions makes sure the number of benchmark runs is right with the configuration"""
```
Comment (kasperski95) [inline tests/performance/statistical_test.py:54]:
```suggestion
def assert_benchmark_runs_matches_config(run):
```

File: tests/performance/statistical_test.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import json
```
Comment (kasperski95) [inline tests/performance/statistical_test.py:1]: Types for arguments would be useful.

File: tests/performance/test_performance_testing.py (hunk 2 above)
Comment (Arcticae) [reply tests/performance/test_performance_testing.py:144]: asyncio, a common abbreviation

File: tests/performance/test_performance_testing.py (hunk 3 above)
Comment (Arcticae) [reply tests/performance/test_performance_testing.py:134]: yep, kind of. helpful with debugging and it doesn't really hurt

File: tests/performance/test_performance_testing.py [hunk 6]
```diff
@@ -97,44 +102,112 @@ def build_test_suite(
...
+            await runner._invoke_cases(test_suite, execution_state)
 
-            print(test_result.format(), file=out)
-            tests_left -= 1
+        yield run
 
-        assert (
-            not tests_state.any_failed_or_broken()
-        ), "Tests failed! See logs above for more info"
+        wait_for_completion(test_suite, shared_state)
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply tests/performance/test_performance_testing.py:206]: this is out of scope, can we prepare another PR for this?

File: tests/performance/test_performance_testing.py (hunk 1 above)
Comment (Arcticae) [reply tests/performance/test_performance_testing.py:210]: oh thx!

File: pyproject.toml [hunk 7]
```diff
@@ -60,7 +60,7 @@ test_cairo = "python ./binary_entrypoint.py test"
 test_e2e.shell = "pytest tests/e2e"
 test_html = [{shell = "coverage html && open ./htmlcov/index.html"}]
 test_integration.shell = "coverage run -a -m pytest tests/integration"
-test_performance = "coverage run -a -m pytest tests/performance --benchmark-compare --benchmark-compare-fail=mean:10%"
+test_performance.shell = "coverage run -a -m pytest tests/performance --benchmark-save-data --benchmark-autosave" #  && python ./tests/performance/statistical_test.py
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply pyproject.toml:63]: the purpose of those tests isn't increasing coverage, we'll probably be running them separately from cov anyways

File: protostar/commands/test/test_runner.py [hunk 8]
```diff
@@ -112,11 +121,11 @@ async def run_test_suite(
                 )
             )
 
-    async def _run_test_suite(
+    async def _prepare_suite(
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/commands/test/test_runner.py]: i'll agree to that but i have to say that naming in protostar is not really intuitive, sometimes making things more verbose actually does not help with the understanding of the code

File: tests/performance/test_performance_testing.py (hunk 3 above)
Comment (Arcticae) [reply tests/performance/test_performance_testing.py:134]: actually scratch that. i will delete those

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply pyproject.toml:64]: `coverage run -a -m ` should be removed then

File: protostar/commands/test/test_runner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/test_runner.py]: I don't see how making things less verbose would make the code easier to understand.

> Explicit is better than implicit.
https://en.wikipedia.org/wiki/Zen_of_Python

**Review by mkaput — APPROVED** [review]

**Pull request merged**